### PR TITLE
Add actions for integration manager to send and read certain events

### PR DESCRIPTION
**Pull request by justinbot**
Adds new `send_event` and `read_events` actions for the integration manager API like those in the Widget API.
This enables integration manager functionality to be expanded in a generic way, while improving parity with the Widget API which it should eventually use instead (https://github.com/vector-im/element-meta/issues/792).

### Implementation
These new actions are implemented as described for the Widget API in https://github.com/matrix-org/matrix-spec-proposals/pull/2762 (except for the limitations that follow).

### Security considerations
MSC2762 describes important [security considerations](https://github.com/matrix-org/matrix-spec-proposals/blob/travis/msc/widgets-send-receive-events/proposals/2762-widget-event-receiving.md#security-considerations) for this API, which are addressed in this implementation by some limitations:

First, this API can only send and read state events -- sending and receiving message events is not implemented.

Second, this API can only send and read certain types of events.
In the Widget API this is handled by capabilities negotiation and prompting the user for permission. However the integration manager API has no such functionality, so a handful events which are relevant to integration managers have been specified:
* Can read:
  * `m.room.power_levels`
  * `m.room.encryption`
  * `m.room.member`
  * `m.room.name`
  * `m.widgets`
  * `im.vector.modular.widgets`
  * `io.element.integrations.installations`
* Can send:
  * `m.widgets`
  * `im.vector.modular.widgets`
  * `io.element.integrations.installations`

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add actions for integration manager to send and read certain events ([\#9740](https://github.com/matrix-org/matrix-react-sdk/pull/9740)).

### CODE REVIEW

File: src/ScalarMessaging.ts [hunk 1]
```diff
@@ -693,6 +720,141 @@ async function getOpenIdToken(event: MessageEvent<any>) {
     }
 }
 
+async function sendEvent(
+    event: MessageEvent<{
+        type: string,
+        state_key?: string,
+        content?: unknown,
```
Comment (SimonBrandner) [inline src/ScalarMessaging.ts:727]:
```suggestion
        content?: IContent,
```

Perhaps this would be better?

File: src/ScalarMessaging.ts [hunk 2]
```diff
@@ -693,6 +720,141 @@ async function getOpenIdToken(event: MessageEvent<any>) {
...
+    if (stateKey !== undefined) {
+        // state event
+        try {
+            res = await client.sendStateEvent(roomId, eventType, content, stateKey);
+        } catch (e) {
+            sendError(event, _t("Failed to send event"), e);
+        }
+    } else {
+        // message event
+        sendError(event, _t("Failed to send event"), new Error("Not implemented"));
```
Comment (SimonBrandner) [inline src/ScalarMessaging.ts:776]:
```suggestion
        sendError(event, _t("Failed to send event"), new Error("Sending non-state events not implemented"));
```

File: src/ScalarMessaging.ts [hunk 3]
```diff
@@ -693,6 +720,141 @@ async function getOpenIdToken(event: MessageEvent<any>) {
...
+        const effectiveStateKey = stateKey === true ? undefined : stateKey;
+        const events: MatrixEvent[] = [].concat(room.currentState.getStateEvents(eventType, effectiveStateKey) || []);
+
+        sendResponse(event, {
+            events: events.slice(0, effectiveLimit).map(e => e.getEffectiveEvent()),
+        });
+        return;
+    } else {
+        // message events
+        sendError(event, _t("Failed to read events"), new Error("Not implemented"));
```
Comment (SimonBrandner) [inline src/ScalarMessaging.ts:853]:
```suggestion
        sendError(event, _t("Failed to read events"), new Error("Reading non-state events implemented"));
```

**Review by SimonBrandner — APPROVED** [review]

**Comment by SimonBrandner** [comment]
> Looks like all checks passed except SonarCloud (which is not expected to pass because this is covered by E2E tests)

Yeah, I think you need to add Jest tests or get someone with force-merge rights to merge this

**Review by turt2live — APPROVED** [review]

**Comment by turt2live** [comment]
err, it should be, yes. For force-merge requests, we generally use the Web App Team internal room as I am by far not the best person on the team to pick this stuff up 😅

**Pull request merged**